pkg/reconciler/reconciler.go

```diff
@@ -96,8 +97,15 @@ func NewBase(opt Options, controllerAgentName string) *Base {
 	eventBroadcaster := record.NewBroadcaster()
 	eventBroadcaster.StartLogging(logger.Named("event-broadcaster").Infof)
 	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: opt.KubeClientSet.CoreV1().Events("")})
-	recorder := eventBroadcaster.NewRecorder(
-		scheme.Scheme, corev1.EventSource{Component: controllerAgentName})
+
+	// Use recorder provided in options if presents.   Otherwise, create a new one.
+	var recorder record.EventRecorder
+	if opt.Recorder == nil {
+		recorder = eventBroadcaster.NewRecorder(
+			scheme.Scheme, corev1.EventSource{Component: controllerAgentName})
+	} else {
```

Comment:
Thanks for your great comment!  The idea is to allow caller to provide recorder (or fake recorder)  as an option.   It will fall back to the default behavior (namely, creating broadcast/recorder internally by reconciler (when opt.Recorder is nil).   Does it make sense?  I agree that the record.NewBroadcaster() should probably not be created when recorder is provided.   Perhaps, something like the following?
```
	if opt.Recorder == nil {

		// Create event broadcaster
		logger.Debug("Creating event broadcaster")
		eventBroadcaster := record.NewBroadcaster()
		eventBroadcaster.StartLogging(logger.Named("event-broadcaster").Infof)
		eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: opt.KubeClientSet.CoreV1().Events("")})

		recorder = eventBroadcaster.NewRecorder(
			scheme.Scheme, corev1.EventSource{Component: controllerAgentName})
	} else {
		logger.Debug("Using recorder from option")
		recorder = opt.Recorder
	}
```

Reply:
Yes I like this much more. It makes the distinction very clear about using `opt.Recorder`.

Reply:
maybe

```go
recorder = opt.Recorder
if recorder == nil {
  // […] 
}
```

Reply:
Sure.